src/usage_metrics/raw/extract.py

```diff
@@ -0,0 +1,100 @@
+"""Generic extraction functionality for data from GCS."""
+
+import os
+from abc import ABC, abstractmethod
+from pathlib import Path
+from tempfile import TemporaryDirectory
+
+import pandas as pd
+from dagster import (
+    AssetExecutionContext,
+)
+from google.api_core.page_iterator import HTTPIterator
+from google.cloud import storage
+from tqdm import tqdm
+
+
+class GCSExtractor(ABC):
+    """Generic extractor base class for Google Cloud Storage logs."""
+
+    def __init__(self, *args, **kwargs):
+        """Create new extractor object and load metadata.
+
+        Args:
+            ds (datastore.Datastore): An initialized datastore, or subclass
+        """
+        if not self.dataset_name:
+            raise NotImplementedError("self.dataset_name must be set.")
+        if not self.bucket_name:
+            raise NotImplementedError("self.bucket_name must be set.")
+
+    @abstractmethod
+    def filter_blobs(
+        self, context: AssetExecutionContext, blobs: HTTPIterator
```

Comment:
Is there a way to design this so we don't need access to dagster context?

Reply:
I can imagine an entirely non-partitioned asset which would not need `partition_key` as an input to `filter_blobs`, so I'm hesitant to hardcode `partition_key` as an input for the base class. Which would mean needing to subclass the `extract` method for each datasource in order to pass in the correct kwargs, unless I'm missing something obvious?

Reply:
Ah good point about`extract()`. Let's keep it as is. Building a context object isn't so bad.